src/java/org/apache/cassandra/index/sai/disk/v1/segment/SegmentBuilder.java

```diff
@@ -112,54 +114,31 @@ protected SegmentMetadata.ComponentMetadataMap flushInternal(IndexDescriptor ind
                                                                indexIdentifier,
                                                                indexTermType.fixedSizeOf(),
                                                                maxSegmentRowId);
-            return writer.writeCompleteSegment(trieBuffer.iterator());
+            return writer.writeCompleteSegment(segmentRamBuffer.iterator());
         }
     }
 
-    public static class RAMStringSegmentBuilder extends SegmentBuilder
+    public static class LiteralSegmentBuilder extends RamBufferSegmentBuilder
     {
-        final RAMStringIndexer ramIndexer;
-
-        final BytesRefBuilder stringBuffer = new BytesRefBuilder();
-
-        public RAMStringSegmentBuilder(IndexTermType indexTermType, NamedMemoryLimiter limiter)
+        public LiteralSegmentBuilder(IndexTermType indexTermType, NamedMemoryLimiter limiter)
         {
             super(indexTermType, limiter);
-
-            ramIndexer = new RAMStringIndexer();
-            totalBytesAllocated = ramIndexer.estimatedBytesUsed();
-        }
-
-        @Override
-        public boolean isEmpty()
-        {
-            return ramIndexer.isEmpty();
         }
 
         @Override
         protected long addInternal(ByteBuffer term, int segmentRowId)
         {
-            copyBufferToBytesRef(term, stringBuffer);
-            return ramIndexer.add(stringBuffer.get(), segmentRowId);
+            return segmentRamBuffer.add(v -> ByteSource.fixedLength(term), term.limit(), segmentRowId);
```

Comment:
Do all literal terms have fixed length?

Reply:
I'm going to have a look at this. I'm not sure why they need a special encoding here and in the in-memory index. Atm, if I change it, tests fail and that doesn't feel right. I just haven't had a chance to have a close look at it.